gl: Require ARB_shader_bit_encoding in FS

gerddie opened the pull request:
On drivers only supportinmg OpenGL 3.3, like llvmpipe, ARB_gpu_shader5 may
not be available, but (u)intBitsToFloat and floatBitsTo* functions are also
provided by ARB_shader_bit_encoding. By requiring this extension in the FS
shader traces that were captured with the llvmpipe driver can actually also
be replayed by using this driver.

baldurk commented:
Can you take a step back and explain exactly what the root problem is? ARB_gpu_shader5 is not a required extension, so it's not a problem to run on a driver that doesn't support it. On those implementations renderdoc will use a fallback path and pick integer pixel values through storing their values in floats, which is accurate up to 16-bit integer values at least.

gerddie commented:
I've opened issue #1586 to explain the issue in more detail and also reworded the commit message. The key point is that `(u)intBitsToFloat` is actually part of GLSL 330, but  you set the version to 150. So a shader that contains `(u)intBitsToFloat` needs something to enable this functionality, and ARB_shader_bit_encoding is this little something. 

An alternative solution would be to set the language level based on the shader code, (i.e. the failing shader actually sets `#version 330`). If you prefer I can also look into this instead. 

PS: regarding `clang-format-3.8`: I understand your argument,  but llvm/clang 3.8  is quite old, and no longer available in most distros.

baldurk commented:
I've continued the discussion in #1586. This fix seems like it might be hiding the real problem. I understand how extension enabling works in GL and that these functions were not core in GLSL 150, but it seems like something else is going on because there is already a check to protect that the functions are only used when the `ARB_gpu_shader5` extension is available.

As it stands a correctly behaving compiler shouldn't do anything different with this change because the functionality will still be guarded behind a check for a different extension. I don't want to change the language version used because it's less of a headache using a minimum version + conditional extensions, rather than trying to reliably figure out the highest language version supported.

Yes it's a bit annoying that I have to fix a clang-format version, but without clang-format output remaining stable over versions I'd have to completely reformat the whole repository with a new version and force all contributors to update all in lockstep. It would be constantly tail-chasing and it's more important that everyone has a uniform version to run. LLVM provides binary releases of 3.8.1 that can be downloaded and run on any distro as they have minimal dependencies.

baldurk commented on renderdoc/driver/gl/gl_debug.cpp:
```diff
@@ -410,6 +410,15 @@ void GLReplay::InitDebugData()
+    else if(HasExt[ARB_shader_bit_encoding])
+      texSampleDefines +=
+          "#define HAS_BIT_CONVERSION 1\n"
+          "#extension GL_ARB_shader_bit_encoding : require\n";
```
This means the OpenGL ES path won't work at all, `HAS_BIT_CONVERSION` should be defined unconditionally in the `if(IsGLES)` branch above because we require version 310 at minimum which supports the functions we want.

gerddie replied:
Define added to the GLES code path.

baldurk reviewed (changes requested):
Thanks this looks OK now I think, aside from a couple of tweaks. I left a note above, and I'd also ask you to reword the commit message since it's misleading to the point of being incorrect:

> Since enabling ARB_gpu_shader5 unconditionally in the shader code seem to let glslang enable the define GL_ARB_gpu_shader5 add it only to the shader if it is really available.

ARB_gpu_shader5 wasn't enabled unconditionally in the shader code, the core problem was that the shader code was set up to do "enable ARB_gpu_shader5 if available, then use the preprocessor to have a fallback path if not". However glslang's preprocessor is entirely independent of the driver so when it preprocesses it bases extension availability on whether each extension is supported by glslang - which in this case it is - and strips out the preprocessor fallback before it reaches the driver and allows the driver to use the fallback path.

Using our own external macro define `HAS_BIT_CONVERSION` allows us to control that first preprocessor pass and ensure things are consistent with actual extension availability. That is the core bugfix, adding any further fallback to use `ARB_shader_bit_encoding` when `ARB_gpu_shader5` isn't available is a tangential middle fallback path (between full `ARB_gpu_shader5` support and nothing). E.g. if a driver didn't support either extension we'd still need the `HAS_BIT_CONVERSION` fix to hit the fallback path correctly.

gerddie committed:
gl: Handle the use of (u)intBitsToFloat by using a separate define

The shader code is preprocessed with glslang that doesn't use the driver
capabilities to define macros indicating the availability of extensions,
instead it uses its own set of supported extensions, and in this case
it supports ARB_gpu_shader5. As a result the generated shader code would
make use of the (u)intBitsToFloat conversion functions based on the define
GL_ARB_gpu_shader5 also if the driver doesn't support this extension.

Hence an additional define is added to indicate whether bit casting
functions are available, and the extension is only enabled/required when
the driver actually supports it.

Closes: #1586

v2: - Fix ws
    - reword commit message

v3: - check for ARB_shader_bit_encoding and ARB_gpu_shader5 in shader creation code
    - don't enable GL_ARB_gpu_shader5 unconditionally in the OpenGL shader
    - reword commit message

v4: Use an additional define HAS_BIT_CONVERSION to make sure gslang correctly
    parses the shader code (Baldurk).

v6: - Add HAS_BIT_CONVERSION also with the GLES code path (Baldur)
    - split patch
    - reword commit message(s) (Baldur)

v7: fix ws

Signed-off-by: Gert Wollny <gert.wollny@collabora.com>

lala

gerddie committed:
gl: Consider also ARB_shader_bit_encoding to enable bit conversion

On drivers only supportinmg OpenGL 3.3, like llvmpipe, ARB_gpu_shader5 may
not be available, but (u)intBitsToFloat and floatBitsTo* are part of GLSL 330,
and are also availabe by enabling the extension ARB_shader_bit_encoding.

Consequently, check for this extension when ARB_gpu_shader5 is not available,
require it when available, and enable the bit code path using these bit cast
functions accordingly.

Signed-off-by: Gert Wollny <gert.wollny@collabora.com>

gerddie commented:
I've split the patch into two to separate the change to handle ARB_gpu_shaders5 and adding the fallback with ARB_shader_bit_encoding.

baldurk reviewed (approved):
LGTM now, thanks. Once CI is green I'll merge.
